azurelinuxagent/ga/monitor.py

```diff
@@ -269,7 +268,12 @@ def collect_and_send_events(self):
 
             self.last_event_collection = datetime.datetime.utcnow()
 
-    def daemon(self):
+    def daemon(self, init_data=False):
+
+        if init_data:
```

Comment:
Now initializing the protocol as part of the new thread.
`init_sysinfo` uses certain properties from the protocol object due to which it was moved here too

Reply:
So we were only doing it once in the earlier code, as part of `monitor_thread.run()` and in the main loop we were only calling `monitor_thread.start()`
```python
    def run(self):
        self.init_protocols()
        self.init_sysinfo()
        self.start()

            while self.running:
                #
                # Check that the parent process (the agent's daemon) is still running
                #
                if not debug and self._is_orphaned:
                    logger.info("Agent {0} is an orphan -- exiting", CURRENT_AGENT)
                    break

                #
                # Check that all the threads are still running
                #
                if not monitor_thread.is_alive():
                    logger.warn(u"Monitor thread died, restarting")
                    monitor_thread.start()
```

I guess they made it that way to not re-run the `init_sysinfo` and `init_protocol` to make sure some basic properties are not overwritten. With my changes, I did not want to modify that behavior so I added the flag `init_data` to ensure that we're only initializing the data once and not everytime the thread goes down.

If you want to discuss about this more we can do it offline

Reply:
Got it, thanks!